lms/djangoapps/instructor/lti_grader.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python
+"""
+Send grades to an OpenEdX LTI component
+"""
+import csv
+
+from instructor import lti_connection
+from instructor.lti_connection import LTIConnectionError
+
+
+class LTIGrader(object):
+    """
+    Updates grades for a given LTI component
+    """
+
+    def __init__(self, course_id, url_base, lti_key, lti_secret):
+        self.url_base = url_base
+        self.course_id = course_id
+        self.key = lti_key
+        self.secret = lti_secret
+        self.output = {'success': '', 'error': ''}
+
+    def _get_first_anon_id(self, grade_data):
+        """
+        Returns the first anonymized student id for in the given uploaded csv data file
+        :param grade_data: an InMemoryUploadedFile with lti grade data
+        :return: anon_id
+        """
+        reader = self._unicode_csv_reader(grade_data)
+        try:
+            reader.next()
+            first_row = reader.next()
+            anon_id = first_row[1]
+        except StopIteration:
+            self.output['error'] += '\nThe CSV file contains no useful data!'
+            anon_id = None
+        return anon_id
+
+    def update_grades(self, grade_data):
+        """
+        Updates grades for the LTI component with the grade data provided
+        :param grade_data: an InMemoryUploadedFile with lti grade data
+        :return: output, a dictionary with two keys: 'success' and 'error'
+        """
+        grade_data = grade_data.read().splitlines()
+
+        first_anon_id = self._get_first_anon_id(grade_data)
+        if first_anon_id is None:
+            return self.output
+
+        test_url = self.url_base + first_anon_id
+        try:
+            lti_connection.validate_lti_passport(self.key, self.secret, test_url)
```

Comment:
Since the anonymized ids are generated by the system and one is required to run the passport sanity check, we take the first one. The reason for running one passport validation once before beginning to process grades, is so that in the case that the user has invalid lti credentials, we do not attempt to post grades for each of the rows and simply inform the user that their credentials were invalid.